### PR TITLE
Exp webgl

### PR DESCRIPTION
**Pull request by htoo97**

### CODE REVIEW

File: phyloviz/static/callbacks.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+		//selective render --  when zooming out
+		/*if(window.zoomLevel % 5 === 0) {
+			console.log("zoom collapse");
+		}*/
+	}
+}
+
+/*
+TODO: need to change edge metadata to hold color as "#RRGGBB" hex string and the enable webgl to parse it
```
Comment (mortonjt) [inline phyloviz/static/callbacks.js]: We'll want to push all of these TODOs to the issues.  More detail on the issues will be necessary.

File: phyloviz/controller.py [hunk 2]
```diff
@@ -1,83 +1,74 @@
...
+sys.path.append("../..")
+import numpy as np
+import pandas as pd
+from skbio import DistanceMatrix, TreeNode
+from scipy.cluster.hierarchy import ward, complete
+
+# trees
 tree = model.read('./astral.MR.rooted.nid.nosup.nwk', 'newick')
+# tree = model.read('./gg_13_5_otus_99_annotated.tree', 'newick')
+# tree = model.read('./0B5tlRtQ-tBfkZSuneOKbg.nwk', 'newick')
```
Comment (mortonjt) [inline phyloviz/controller.py]: We will want to have a more versatile way to load in these files.  For starters, let's wrap this in functions so that the trees / metadata can be directly passed in.

File: phyloviz/test.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+from model import Model
+from phyloviz.model import Tree
+import pandas as pd
+
+tree = model.read('./astral.MR.rooted.nid.nosup.nwk', 'newick')
+m = Model(tree,'ncbi.t2t.txt','metadata.txt')
+edgeM = m.retrive_view_coords()
+#print(edgeM)
+edge = m.updateEdgeCategory('marker_count','width',0.5,lower=300,upper=500)
+print(edge)
```
Comment (mortonjt) [inline phyloviz/test.py]: Its great that there are helper scripts here, but we'll want to think about more robust ways of testing via unittests.

**Comment by mortonjt** [comment]
@htoo97 @kwcantrell @YimengYang this is a lot of code!

It's getting to the point where we will want to start drawing UML diagrams to determine which functions are connected.  Have you guys drawn these before?

@ElDeveloper @wasade feel free to chime in on this.

File: phyloviz/static/callbacks.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+/*
+ * Collapses clades based on the slider scale
+ */
+function collapseClades() {
+	var ss = document.getElementById("collapseRange").value;
+	console.log('collapse clade', ss);
+
+	var triangles;
+	var edges;
+	$.getJSON('http://localhost:8080/collapse', {sliderScale: ss}, function(data) {
```
Comment (mortonjt) [inline phyloviz/static/callbacks.js]: We will want to have this be a variable - this will be especially useful as we shuttle over different urls / ports.

File: phyloviz/static/callbacks.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+		triangles = data;
+	}).done(function() {
+		$.getJSON('http://localhost:8080/api/edges', function(data2) {
```
Comment (mortonjt) [inline phyloviz/static/callbacks.js]: Same here.

File: phyloviz/controller.py (hunk 2 above)
Comment (YimengYang) [reply phyloviz/controller.py]: Addressed.

File: phyloviz/model.py [hunk 6]
```diff
@@ -2,19 +2,54 @@
...
+    -------
+       pd.Dataframe
+
+    """
+    metadata = pd.read_table(file_name)
+    metadata.rename(columns={metadata.columns[0]: "Node_id"}, inplace=True)
+    return metadata
+
+
+def read_internal_node_metadata(file_name):
```
[the remark this answers is not in the document]
Comment (YimengYang) [reply phyloviz/model.py]: It might need to change based on the file format.

File: phyloviz/static/callbacks.js [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+	else if(event.deltaY < 0) {
+		var scaleByMat = new Float32Array(16);
+		var scaleAmount = 1 / window.scaleFactor;
+		window.zoomAmount = window.zoomAmount / scaleAmount;
+		var scaleFactorVec = vec3.fromValues(scaleAmount, scaleAmount, scaleAmount);
+		mat4.fromScaling(scaleByMat, scaleFactorVec);
+		mat4.mul(window.worldMat, scaleByMat, window.worldMat);
+		window.zoomLevel--;
+
+		//selective render --  when zooming out
```
[the remark this answers is not in the document]
Comment (YimengYang) [reply phyloviz/static/callbacks.js]: Fixed.

File: phyloviz/static/callbacks.js (hunk 4 above)
Comment (YimengYang) [reply phyloviz/static/callbacks.js]: Addressed.

File: phyloviz/static/callbacks.js (hunk 5 above)
Comment (YimengYang) [reply phyloviz/static/callbacks.js]: Addressed.

File: phyloviz/static/callbacks.js [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+function selectColor() {
+	console.log('Color option selected');
+	var edges;
+	var attr = document.getElementById("metadataOptions").value;
+	var l = document.getElementById("lowerBound").value;
+	var u = document.getElementById("upperBound").value;
+	var e = document.getElementById("category").value;
+	$.getJSON('http://localhost:8080/color', {attribute : attr, lower : l, equal : e, upper : u}, function(data) {
```
[the remark this answers is not in the document]
Comment (YimengYang) [reply phyloviz/static/callbacks.js]: Addressed.

File: phyloviz/tree_with_webgl.html [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+		.grid {
+			display: grid;
+			grid-template-columns: 3fr 1fr;
+			grid-column-gap: 4px;
+			grid-auto-rows: auto;
+			height: 98vh;
+		}
+
+		.metadataTabs {
```
Comment (qiyunzhu) [inline phyloviz/tree_with_webgl.html]: I have the impression that, web developers typically use names like `metadataTabs` for JavaScripts variables, but `metadata-tabs` for CSS classes.

File: phyloviz/tree_with_webgl.html [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+			    		<div class="textField">
+				    		<label for="category">Category Name</label>
+				    		<input type="text" id="category">
+			    		</div>
+		    		</div>
+
+		    		<!-- show color menu and hide others-->
+		    		<div class="menu" id="showColorMenu">
+						<div>
+				 	 		<select id = "metadataOptions"></select>
```
Comment (qiyunzhu) [inline phyloviz/tree_with_webgl.html]: Extra spaces.

File: phyloviz/tree_with_webgl.html [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+		<script>
+			var edgeMetadata;
+			$.getJSON('http://localhost:8080/api/edges', function(data) {
+				edgeMetadata = data; // JSON.parse(JSON.stringify(data));
+			}).done(function() {
+				console.log(edgeMetadata);
+				InitWebGl(edgeMetadata);
+			});
+
+            var slider = document.getElementById("collapseRange");
```
Comment (qiyunzhu) [inline phyloviz/tree_with_webgl.html]: Why stop using JQuery here?

File: phyloviz/webserver.py [hunk 12]
```diff
@@ -1,21 +1,27 @@
 import tornado
-from controller import ModelHandler, IndexHandler, NodeHandler, EdgeHandler
+from controller import ModelHandler, EdgeHandler, ZoomHandler
```
Comment (qiyunzhu) [inline phyloviz/webserver.py]: These can be merged as:
```
from controller import (
    ModelHandler, EdgeHandler, ZoomHandler,
    BenchmarkHandler, HighlightHandler, CollapseHandler)
```

File: phyloviz/model.py [hunk 13]
```diff
@@ -214,30 +200,48 @@ def coords(self, height, width):
         """
 
         # calculates coordinates of all nodes
-        self.rescale(width, height)
+        print("start")
+        start = time.time()
+        scale = self.rescale(width, height)
+        print(time.time() - start)
```
Comment (qiyunzhu) [inline phyloviz/model.py]: Maybe some time formatting will make the output more readable.

File: phyloviz/model.py [hunk 14]
```diff
@@ -214,30 +200,48 @@ def coords(self, height, width):
...
-            nodeData[node.name] = {**nId, **coords}
+            attr = {'color_R': 255.0, 'color_G': 255.0, 'color_B': 255.0,
+                    'is_visible': True, 'size': 1}
+            nodeData[node.name] = {**nId, **coords, **attr}
 
         # edge metadata
         edgeData = {}
         for node in self.postorder():
-            pId = {'Parent id': node.name}
+            if node.is_tip():
```
Comment (qiyunzhu) [inline phyloviz/model.py]: This can be:
```
edgeData["is_tip"] = node.is_tip()
```

File: phyloviz/model.py [hunk 15]
```diff
@@ -327,9 +330,10 @@ def update_coordinates(self, s, x1, y1, a, da):
             y1 = node.parent.y2
             a = node.parent.angle
 
-            # calculates 'a'
             a = a - node.parent.leafcount * da / 2
```
Comment (qiyunzhu) [inline phyloviz/model.py]: In Python there are also `+=` and `-=`, like many other languages.

File: phyloviz/model.py [hunk 16]
```diff
@@ -327,9 +330,10 @@ def update_coordinates(self, s, x1, y1, a, da):
...
             for sib in node.parent.children:
+                if len(node.parent.children) < 2:
```
Comment (qiyunzhu) [inline phyloviz/model.py]: I don't understand the logic here. This mean single leaf, right?

File: phyloviz/model.py [hunk 17]
```diff
@@ -341,16 +345,24 @@ def update_coordinates(self, s, x1, y1, a, da):
             (node.x1, node.y1, node.x2, node.y2, node.angle) = (x1, y1, x2,
                                                                 y2, a)
 
-            # TODO: Add functionality that allows for collapsing of nodes
-            if node.is_tip():
-                points += [(x2, y2)]
+            if x2 > max_x:
```
Comment (qiyunzhu) [inline phyloviz/model.py]: I would do:
```
max_x, min_x = max(max_x, x2), min(min_x, x2)
```

File: phyloviz/model.py [hunk 18]
```diff
@@ -450,232 +443,200 @@ def layout(self, layout_type):
...
-        Based on the zoom level, coords will be recalculated with resolution
-        resolving.
-        This function should also call mask on _canvascoords to recalculate the
-        coords that get displayed.
-
-        This function will be called by zoom.
-
-        Pipeline function
-
+    def uniqueCategories(metadata, attribute):
```
Comment (qiyunzhu) [inline phyloviz/model.py]: I think a more Pythonic function name would be `unique_categories`.

File: phyloviz/model.py [hunk 19]
```diff
@@ -450,232 +443,200 @@ def layout(self, layout_type):
...
-
-
-        # TODO:
-        # what should this transform function do??
-        # 1) Should it update the mask?
-        It needs to call mask by passing in the zoom point
-        # 2) Should it only handle handling?
-        # 3) Perspective transform (i.e. create fish bowl effect)
-        # are we missing anything?
+        unique_cat : list
```
Comment (qiyunzhu) [inline phyloviz/model.py]: Should say something like `list of str`.

File: phyloviz/static/callbacks.js [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+*/
+
+/*
+ * Highlights the user selected metadata
+ */
+function selectHighlight() {
+	console.log('Highlight option selected');
+	var edges;
+	var attr = document.getElementById("metadataOptions").value;
+	var l = document.getElementById("lowerBound").value;
```
Comment (qiyunzhu) [inline phyloviz/static/callbacks.js]: Since you already used jQuery, you should just use it in all instances. For example:
native JavaScript:
```
document.getElementById('demo').innerHTML = 'Hello JavaScript';
```
jQuery:
```
$('#demo').html('Hello jQuery');
```

File: phyloviz/benchmark.html [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+			stroke-width: 2px;
+		}
+		.link {
+			stroke: #777;
+			stroke-width: 2px;
+		}
+	</style>
+</head>
+
+<svg width="1000" height="700"></svg>
```
[the remark this answers is not in the document]
Comment (kwcantrell) [reply phyloviz/benchmark.html]: The most recent code has a dynamic size. I redid the web interface to use a grid layout along with a css style sheet so it will automatically resize according to the users screen..

File: phyloviz/benchmark.html [hunk 22]
```diff
@@ -0,0 +1,167 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+	<meta charset="utf-8">
+	<title>Tree Example</title>
+
+
+		<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js" type="text/javascript"></script>
```
Comment (ElDeveloper) [inline phyloviz/benchmark.html]: I would recommend that all the resources are loaded locally, and that the versions and licenses are pinned. If not at this moment, maybe down the road.

File: phyloviz/benchmark.html [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+		var d3_level = 1;
+		var trans_x = 0;
+		var trans_y = 0;
+
+
+		/*document.body.style.width = '700px';
+		document.body.style.height = '1000px';*/
+
+		$.getJSON('http://localhost:8080/api/edges', function(data) {
+			edges = JSON.parse(JSON.stringify(data));
```
Comment (ElDeveloper) [inline phyloviz/benchmark.html]: Why does the data need to be converted into a JSON string and parsed if it's already being requested as a JSON string?

File: phyloviz/benchmark.html [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+			stroke-width: 2px;
+		}
+	</style>
+</head>
+
+<svg width="1000" height="700"></svg>
+
+<body>
+	<script>
+		//<svg width="1000" height="700"></svg> Was placed above <body>
```
Comment (ElDeveloper) [inline phyloviz/benchmark.html]: There's a number of blocks in this tag that are commented, I suggest to remove them. If they are needed "later" you can always go back using git.

File: phyloviz/benchmark.html (hunk 23 above)
Comment (ElDeveloper) [reply phyloviz/benchmark.html]: Whatever applies to this block should apply to all the other occurrences of this pattern throughout this PR.

File: phyloviz/tree_with_webgl.html (hunk 9 above)
Comment (ElDeveloper) [reply phyloviz/tree_with_webgl.html]: 👍 on using hyphen-separated names.

File: phyloviz/static/init_webgl.js [hunk 25]
```diff
@@ -0,0 +1,211 @@
...
+ * Extracts the coordinates of the tree from edge_metadata
+ */
+function extractEdges(edgeMeta) {
+	var minX = Infinity;
+	var maxX = -Infinity;
+	var minY = Infinity;
+	var maxY = -Infinity;
+	window.result = [];
+	window.edgeMetadata = edgeMeta;
+	for(i = 0; i < window.edgeMetadata.length; i++){
```
Comment (ElDeveloper) [inline phyloviz/static/init_webgl.js]: can you declare `i` so it doesn't leak to the global scope?

File: phyloviz/static/drawing_loop.js [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+	//
+	// Main render loop
+	//
+	var loop = function() {
+		window.gl.uniformMatrix4fv(matWorldUniform, window.gl.FALSE, window.worldMat);
+
+		window.gl.clearColor(0.75, 0.85, 0.8, 1.0);
+		window.gl.clear(window.gl.COLOR_BUFFER_BIT | window.gl.DEPTH_BUFFER_BIT);
+		window.gl.drawArrays(window.gl.LINES, 0, window.result.length / 3 );
+		requestAnimationFrame(loop);
```
Comment (ElDeveloper) [inline phyloviz/static/drawing_loop.js]: Just a note. With this loop you will redraw a new frame every possible time, which means that even if the objects have not changed since the last rendering, you will redraw the same frame leading to your CPU burning a lot of unnecessary compute. This is fine for now, but it would be a good idea to start integrating logic in this code-base to allow for selectively rendering a new frame i.e. only when something in the view has changed. I suggest opening an issue about this.

File: phyloviz/0B5tlRtQ-tBfkZSuneOKbg.nwk [hunk 27]
```diff
@@ -0,0 +1 @@
...yota:0.04004[85])Eukaryota:0.02708[41])Eukaryota:0.02636[44])Eukaryota:0.06455[87],Leishmania_major:0.45664)Eukaryota:0.10129[100],Giardia_lamblia:0.55482)Eukaryota:0.57543[100],((Nanoarchaeum_equitans:0.81078,(((Sulfolobus_tokodaii:0.17389,Sulfolobus_solfataricus:0.18962)Sulfolobus:0.33720[100],Aeropyrum_pernix:0.43380)Thermoprotei:0.09462[94],Pyrobaculum_aerophilum:0.55514)Thermoprotei:0.12018[100])Archaea:0.15444[100],((Thermoplasma_volcanium:0.10412,Thermoplasma_acidophilum:0.09785)Thermoplasma:0.66151[100],((((Methanobacterium_thermautotrophicum:0.36583,Methanopyrus_kandleri:0.35331)Euryarchaeota:0.07446[99],(Methanococcus_maripaludis:0.28592,Methanococcus_jannaschii:0.13226)Methanococcales:0.23828[100])Euryarchaeota:0.06284[100],((Pyrococcus_horikoshii:0.02786,Pyrococcus_abyssi:0.02179)Pyrococcus:0.02239[100],Pyrococcus_furiosus:0.02366)Pyrococcus:0.36220[100])Euryarchaeota:0.04469[51],(Archaeoglobus_fulgidus:0.34660,(Halobacterium_sp._NRC-1:0.61597,(Methanosarcina_acetivorans:0.02602,Methanosarcina_mazei:0.03087)Methanosarcina:0.30588[100])Euryarchaeota:0.12801[100])Euryarchaeota:0.10395[100])Euryarchaeota:0.06815[62])Euryarchaeota:0.11833[99])Archaea:0.43325[100]):0.88776);
```
Comment (ElDeveloper) [inline phyloviz/0B5tlRtQ-tBfkZSuneOKbg.nwk]: Is this file needed? If so maybe make the filename more descriptive.

File: phyloviz/static/drawing_loop.js (hunk 26 above)
Comment (mortonjt) [reply phyloviz/static/drawing_loop.js]: Nice catch @ElDeveloper , this will need to be addressed at some point.  Raising issue in #26

File: phyloviz/benchmark.html (hunk 22 above)
Comment (mortonjt) [reply phyloviz/benchmark.html]: 👍 I'd suggest looking at the [support_files](https://github.com/biocore/emperor/tree/new-api/emperor/support_files) in Emperor.  This will be so that we don't necessary need an internet connection in order to view these visualizations.

File: phyloviz/benchmark.html (hunk 23 above)
Comment (mortonjt) [reply phyloviz/benchmark.html]: Also we'll need to make sure that the url / port names are variable.

File: phyloviz/controller.py [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+for i, n in enumerate(tree.postorder(include_self=True)):
+    n.length = 1
+    if not n.is_tip() and n.name is None:
+        n.name = "y%d" % i
+
+m = Model(tree, internal_metadata_file, leaf_metadata_file)
+edgeM = m.retrive_view_coords()
+
+
+class ModelHandler(RequestHandler):
```
Comment (mortonjt) [inline phyloviz/controller.py]: Let's try to create unittests for all of the handlers.

Comment (mortonjt) [reply phyloviz/controller.py]: Maybe worthwhile to look at labman tests

https://github.com/jdereus/labman/tree/master/labman/gui/handlers/process_handlers/test

File: phyloviz/controller.py [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+        self.write(edges)
+        self.finish()
+
+
+class BenchmarkHandler(RequestHandler):
+    def get(self):
+        self.render("benchmark.html")
+
+
+class CollapseHandler(RequestHandler):
```
Comment (mortonjt) [inline phyloviz/controller.py]: Can we move this to a separate PR?

Also can we profile this?  There may be a bug in this.
Recommended profilers

https://github.com/rkern/line_profiler
https://docs.python.org/3/library/profile.html

File: phyloviz/static/init_webgl.js [hunk 30]
```diff
@@ -0,0 +1,211 @@
...
+window.viewMat  = mat4.create();
+window.projMat = mat4.create();
+window.treeNormVec = vec3.create();
+window.identityMat = mat4.create();
+window.scaleFactor = 5.0 / 4.0; //how much the tree grows/shrinks during zoom
+
+/*
+ * compliles shader programs and initializes webgl
+ */
+function InitWebGl() {
```
Comment (mortonjt) [inline phyloviz/static/init_webgl.js]: Don't forget to get rid of all of the tabs.  Our standards is to have 2 spaces instead.

Comment (mortonjt) [reply phyloviz/static/init_webgl.js]: Also recommend looking at jslint to clean up javascript: https://github.com/douglascrockford/JSLint

File: setup.py [hunk 31]
```diff
@@ -0,0 +1,62 @@
...
+
+with open('README.md') as f:
+    long_description = f.read()
+
+# version parsing from __init__ pulled from Flask's setup.py
+# https://github.com/mitsuhiko/flask/blob/master/setup.py
+version = "0.0.1"
+
+setup(
+    name='phyloviz',
```
Comment (mortonjt) [inline setup.py]: rename to Empress!

File: setup.py [hunk 32]
```diff
@@ -0,0 +1,62 @@
...
+    version=version,
+    license='BSD',
+    description=description,
+    long_description=long_description,
+    author="phyloviz development team",
+    author_email="jamietmorton@gmail.com",
```
Comment (mortonjt) [inline setup.py]: Email?

File: setup.py [hunk 33]
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016--, gneiss development team.
```
Comment (mortonjt) [inline setup.py]: Can we replace gneiss with Empress?

File: phyloviz/benchmark.html [hunk 34]
```diff
@@ -0,0 +1,167 @@
+<!DOCTYPE html>
```
Comment (mortonjt) [inline phyloviz/benchmark.html]: Can we remove this html file from this pull request?  I don't think this is needed anymore.

File: phyloviz/controller.py [hunk 35]
```diff
@@ -1,83 +1,74 @@
...
-# # initialize tree with branch length and named internal nodes
-# for i, n in enumerate(tree.postorder(include_self=True)):
-#     n.length = 1
-#     if not n.is_tip():
-#         n.name = "y%d" % i
+# initialize tree with branch length Todo: need to incorperate this into model
+for i, n in enumerate(tree.postorder(include_self=True)):
+    n.length = 1
+    if not n.is_tip() and n.name is None:
+        n.name = "y%d" % i
```
[the remark this answers is not in the document]
Comment (mortonjt) [reply phyloviz/controller.py]: Agreed.  I think it is ok to not have `and n.name is None` to make sure that there are no duplicate ids.

File: phyloviz/test.py [hunk 36]
```diff
@@ -0,0 +1,24 @@
+from phyloviz import model
+from model import Model
+from phyloviz.model import Tree
+import pandas as pd
+
+tree = model.read('./astral.MR.rooted.nid.nosup.nwk', 'newick')
+m = Model(tree,'ncbi.t2t.txt','metadata.txt')
+edgeM = m.retrive_view_coords()
+#print(edgeM)
```
Comment (mortonjt) [inline phyloviz/test.py]: Will want to remove this

File: phyloviz/benchmark.html (hunk 34 above)
Comment (mortonjt) [reply phyloviz/benchmark.html]: Unless this is the example of how to run it -- I think this would be more appropriate for a follow up pull request since this will need a bit of thought on how to handle.  We would like this to be easily reproducible by others.

File: phyloviz/controller.py [hunk 37]
```diff
@@ -1,83 +1,93 @@
...
 from tornado.web import RequestHandler
+import sys
+sys.path.append("../..")
+import numpy as np
+import pandas as pd
+from skbio import DistanceMatrix, TreeNode
+from scipy.cluster.hierarchy import ward, complete
+
+# trees
+tree = model.read('./astral.MR.rooted.nid.nosup.nwk', 'newick')
```
Comment (mortonjt) [inline phyloviz/controller.py]: We really need to have an CLI enabled.  Just opened up #27 so that we can tackle in an upcoming PR.

**Comment by mortonjt** [comment]
Here are some things that would really complete this pull request

1. Documentation on how to run the software in #28 
2. Local downloads of jquery and other required javascript libraries (also remove d3 dependency)
3. Remove commented code
4. Fix the merge conflicts

There are a few things that still need to be resolved (i.e. the lack of a CLI) - but we can follow up in #27 

Once I can run this program on my own machine, I think this is good merge as is.

**Comment by htoo97** [comment]
1. We'll work on the documentation as a README file. 
2. That's already taken care of in another updated branch.
3. On it too.
4. We'll take care of it.

Also, Yimeng will be working on the CLI.

File: phyloviz/static/callbacks.js [hunk 38]
```diff
@@ -0,0 +1,171 @@
...
+	if(event.deltaY > 0){
+		var scaleByMat = new Float32Array(16);
+		var scaleAmount = window.scaleFactor;
+		window.zoomAmount = window.zoomAmount / scaleAmount;
+		var scaleFactorVec = vec3.fromValues(scaleAmount, scaleAmount, scaleAmount);
+		mat4.fromScaling(scaleByMat, scaleFactorVec);
+		mat4.mul(window.worldMat, scaleByMat, window.worldMat);
+		window.zoomLevel++;
+
+		//selective render -- when zooming in
```
[the remark this answers is not in the document]
Comment (YimengYang) [reply phyloviz/static/callbacks.js]: Cleaned

File: phyloviz/static/callbacks.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (YimengYang) [reply phyloviz/static/callbacks.js]: Cleaned

File: phyloviz/controller.py (hunk 35 above)
[the remark this answers is not in the document]
Comment (YimengYang) [reply phyloviz/controller.py]: Tried to rename all internal nodes in both tree and metadata table and it turned out to be really time consuming.

File: phyloviz/tree_with_webgl.html (hunk 9 above)
Comment (YimengYang) [reply phyloviz/tree_with_webgl.html]: Changed

File: phyloviz/tree_with_webgl.html (hunk 10 above)
Comment (YimengYang) [reply phyloviz/tree_with_webgl.html]: Addressed

File: phyloviz/tree_with_webgl.html (hunk 11 above)
Comment (YimengYang) [reply phyloviz/tree_with_webgl.html]: Addressed.

File: phyloviz/webserver.py (hunk 12 above)
Comment (YimengYang) [reply phyloviz/webserver.py]: Addressed.

File: phyloviz/model.py (hunk 14 above)
Comment (YimengYang) [reply phyloviz/model.py]: Changed

File: phyloviz/model.py (hunk 15 above)
Comment (YimengYang) [reply phyloviz/model.py]: Changed

File: phyloviz/model.py (hunk 17 above)
Comment (YimengYang) [reply phyloviz/model.py]: Fixed

File: phyloviz/model.py (hunk 18 above)
Comment (YimengYang) [reply phyloviz/model.py]: Fixed

File: phyloviz/model.py (hunk 19 above)
Comment (YimengYang) [reply phyloviz/model.py]: Fixed

File: phyloviz/static/callbacks.js (hunk 20 above)
Comment (YimengYang) [reply phyloviz/static/callbacks.js]: Cleaned.

File: phyloviz/benchmark.html (hunk 22 above)
Comment (YimengYang) [reply phyloviz/benchmark.html]: Addressed.

File: phyloviz/benchmark.html (hunk 23 above)
Comment (YimengYang) [reply phyloviz/benchmark.html]: Changed.

File: phyloviz/benchmark.html (hunk 24 above)
Comment (YimengYang) [reply phyloviz/benchmark.html]: Removed

File: phyloviz/static/init_webgl.js (hunk 25 above)
Comment (YimengYang) [reply phyloviz/static/init_webgl.js]: Refactored.

File: setup.py (hunk 31 above)
Comment (YimengYang) [reply setup.py]: Renamed

File: setup.py (hunk 32 above)
Comment (YimengYang) [reply setup.py]: Do you mean we should include all of our emails or remove the email?

File: setup.py (hunk 33 above)
Comment (YimengYang) [reply setup.py]: Renamed

File: phyloviz/benchmark.html (hunk 34 above)
Comment (YimengYang) [reply phyloviz/benchmark.html]: Removed

File: README.md [hunk 39]
```diff
@@ -1,2 +1,28 @@
...
+astral.MR.rooted.nid.nosup.nwk, ncbi.t2t.txt, and metadata.txt. "astral.MR.rooted.nid.nosup.nwk" is any Newick tree file.
+"ncbi.t2t.txt" is the metadata file for the internal nodes of the tree. "metadata.txt" is the metadata file for the leaf nodes of the tree.
+
+# Installation
+Please install the following packages in order to run Empress
+
+```
+conda install tornado
+conda install -c conda-forge scikit-bio
+```
```
Comment (mortonjt) [inline README.md]: Maybe we can shorten this to
```
conda create -n empress tornado  scikit-bio -c conda-forge
```

Comment (mortonjt) [reply README.md]: Also, do we need the setup.py working?
